torch/testing/_internal/common_methods_invocations.py

```diff
@@ -1305,6 +1305,69 @@ def sample_inputs_clamp(op_info, device, dtype, requires_grad):
     output += [SampleInput(empty_tensor, args=(0.0, 1.0)), ]
     return output
 
+def sample_inputs_cumprod(op_info, device, dtype, requires_grad):
+    def make_arg(shape):
+        # shrink values to be in the interval [-1, +1] for better precision in gradgradcheck
+        return make_tensor(shape, device, dtype, low=-1, high=+1, requires_grad=requires_grad)
+
+    def prod_zeros(dim_select):
+        assert len(dim_select) == 2
+        result = make_arg(3 * (S,))
+        with torch.no_grad():
+            result.narrow(dim_select[0], 0, 1).narrow(dim_select[1], 1, 1).zero_()
+            result.narrow(dim_select[0], 2, 1).narrow(dim_select[1], 3, 1).zero_()
+            result.narrow(dim_select[0], 4, 1).narrow(dim_select[1], 3, 1).zero_()
+        return result
+
+    # will not be needed once OpInfo tests suport Iterables
+    def sample_generator():
+        for dim in range(3):
+            yield SampleInput((make_arg((S, S, S)), dim))
+        # Scalar tensors and empty tensor
+        for size in [(), (1,), (0,)]:
+            yield SampleInput((make_arg(size), 0))
+
+        yield SampleInput((prod_zeros([0, 1]), 1))
+        yield SampleInput((prod_zeros([0, 2]), 1))
+        yield SampleInput((prod_zeros([1, 2]), 1))
+
+        # test dtype kwarg
+        yield SampleInput((prod_zeros([1, 2]), 1), kwargs={'dtype': dtype})
+
+    return list(sample_generator())
+
+def sample_inputs_prod(op_info, device, dtype, requires_grad):
+    def make_arg(shape):
+        # shrink values to be in the interval [-1, +1] for better precision in gradgradcheck
+        return make_tensor(shape, device, dtype, low=-1, high=+1, requires_grad=requires_grad)
+
+    def prod_single_zero():
+        result = make_arg(2 * (S,))
+        with torch.no_grad():
+            result[0, 1] = 0
+        return result
+
+    # will not be needed once OpInfo tests support Iterables
+    def sample_generator():
+        for sample in sample_inputs_cumprod(op_info, device, dtype, requires_grad):
+            yield SampleInput(sample.input[0])  # only Tensor, ignore other inputs
+            yield sample
+            sample.kwargs['keepdim'] = True
+            yield sample
+        yield SampleInput(prod_single_zero())
```

Comment:
shouldn't we pass the device to `prod_single_zero` here?

Reply:
oh I see you are calling into the function defined above. I think we should remove that and just use `prod_single_zero` defined in `common_utils.py`

Reply:
Only used by a handful of functions like prod/cumprod. I decided it is better to reuse samples.

Reply:
I don't understand. I was suggesting to remove https://github.com/pytorch/pytorch/pull/48125/files#diff-18a0c3e7cd07089b34760028c387d39de1aa393172f1d5f123692106b3a3d144R1344-R1348

Reply:
I am not sure it is very useful to use the version from utils and have it in utils because it is not being used:
```
(pytorch-cuda-dev) nik@qgpu3:~/git/pytorch$ git grep prod_single_zero
torch/testing/_internal/common_methods_invocations.py:    def prod_single_zero():
torch/testing/_internal/common_methods_invocations.py:        yield SampleInput(prod_single_zero())
torch/testing/_internal/common_utils.py:def prod_single_zero(dim_size, device=torch.device('cpu'), dtype=torch.float):
(pytorch-cuda-dev) nik@qgpu3:~/git/pytorch$ git grep prod_zeros
torch/testing/_internal/common_methods_invocations.py:    def prod_zeros(dim_select):
torch/testing/_internal/common_methods_invocations.py:        yield SampleInput((prod_zeros([0, 1]), 1))
torch/testing/_internal/common_methods_invocations.py:        yield SampleInput((prod_zeros([0, 2]), 1))
torch/testing/_internal/common_methods_invocations.py:        yield SampleInput((prod_zeros([1, 2]), 1))
torch/testing/_internal/common_methods_invocations.py:        yield SampleInput((prod_zeros([1, 2]), 1), kwargs={'dtype': dtype})
torch/testing/_internal/common_methods_invocations.py:def prod_zeros(dim_size, dim_select):

```

Reply:
oh I see what you are saying ... `prod_zeros` was only being used for `prod` and `cumprod` method_tests. ok so I think here's what we should do. let's keep the function defined above and let's remove the helper function in `common_utils.py`